File: MiniERP.Server/Mappers/InvoiceMapper.cs
using MiniERP.Server.DTOs;
using MiniERP.Server.Models;

namespace MiniERP.Server.Mappers;

public static class InvoiceMapper
{
    public static InvoiceDTO ToDto(Invoice i) => new()
    {
        Id = i.Id,
        InvoiceNumber = i.InvoiceNumber,
        IssueDate = i.IssueDate,
        TaxableSupplyDate = i.TaxableSupplyDate,
        DueDate = i.DueDate,
        Status = i.Status,
        SupplierId = i.SupplierId,
        CustomerId = i.CustomerId,
        CustomerName = i.Customer?.Name ?? string.Empty,
        TotalAmountExVat = i.TotalAmountExVat,
        VatAmount = i.VatAmount,
        TotalAmountIncVat = i.TotalAmountIncVat,
        CurrencyCode = i.CurrencyCode,
        BankAccount = i.BankAccount,
        VariableSymbol = i.VariableSymbol,
        PaymentMethod = i.PaymentMethod
    };

    public static InvoiceDetailDTO ToDetailDto(Invoice i) => new()
    {
        Id = i.Id,
        InvoiceNumber = i.InvoiceNumber,
        IssueDate = i.IssueDate,
        TaxableSupplyDate = i.TaxableSupplyDate,
        DueDate = i.DueDate,
        Status = i.Status,
        SupplierId = i.SupplierId,
        CustomerId = i.CustomerId,
        TotalAmountExVat = i.TotalAmountExVat,
        VatAmount = i.VatAmount,
        TotalAmountIncVat = i.TotalAmountIncVat,
        CurrencyCode = i.CurrencyCode,
        BankAccount = i.BankAccount,
        VariableSymbol = i.VariableSymbol,
        PaymentMethod = i.PaymentMethod,
        Customer = i.Customer != null ? CustomerMapper.ToDto(i.Customer) : null,
        Supplier = i.Supplier != null ? CustomerMapper.ToDto(i.Supplier) : null,
        Items = i.Items.Select(InvoiceItemMapper.ToDetailDto).ToList()
    };

    public static Invoice ToEntity(CreateInvoiceDTO dto) => new()
    {
        IssueDate = dto.IssueDate,
        TaxableSupplyDate = dto.TaxableSupplyDate,
        DueDate = dto.DueDate,
        CustomerId = dto.CustomerId,
        SupplierId = dto.SupplierId,
        Status = dto.Status,
        TotalAmountExVat = dto.TotalAmountExVat,
        VatAmount = dto.VatAmount,
        TotalAmountIncVat = dto.TotalAmountIncVat,
        CurrencyCode = dto.CurrencyCode,
        BankAccount = dto.BankAccount,
        VariableSymbol = dto.VariableSymbol,
        PaymentMethod = dto.PaymentMethod
    };

    public static void ApplyUpdate(Invoice invoice, UpdateInvoiceDTO dto)
    {
        if (dto.InvoiceNumber != null) invoice.InvoiceNumber = dto.InvoiceNumber;
        if (dto.IssueDate.HasValue) invoice.IssueDate = dto.IssueDate.Value;
        if (dto.TaxableSupplyDate.HasValue) invoice.TaxableSupplyDate = dto.TaxableSupplyDate.Value;
        if (dto.DueDate.HasValue) invoice.DueDate = dto.DueDate.Value;
        if (dto.Status != null) invoice.Status = dto.Status;
        if (dto.TotalAmountExVat.HasValue) invoice.TotalAmountExVat = dto.TotalAmountExVat.Value;
        if (dto.VatAmount.HasValue) invoice.VatAmount = dto.VatAmount.Value;
        if (dto.TotalAmountIncVat.HasValue) invoice.TotalAmountIncVat = dto.TotalAmountIncVat.Value;
        if (dto.CurrencyCode != null) invoice.CurrencyCode = dto.CurrencyCode;
        if (dto.BankAccount != null) invoice.BankAccount = dto.BankAccount;
        if (dto.VariableSymbol != null) invoice.VariableSymbol = dto.VariableSymbol;
        if (dto.PaymentMethod != null) invoice.PaymentMethod = dto.PaymentMethod;
    }
}

File: minierp.client/src/mappers/invoiceMapper.js
/**
 * Mapuje InvoiceDTO na formát vhodný pro editaci
 * @param {Object} invoiceDto
 * @returns {Object}
 */
export const mapInvoiceDtoToForm = (invoiceDto) => {
    return {
        id: invoiceDto.id || "",
        invoiceNumber: invoiceDto.invoiceNumber || "",
        issueDate: invoiceDto.issueDate || "",
        taxableSupplyDate: invoiceDto.taxableSupplyDate || "",
        dueDate: invoiceDto.dueDate || "",
        supplierId: invoiceDto.supplierId || "",
        customerId: invoiceDto.customerId || "",
        totalAmountExVat: invoiceDto.totalAmountExVat || "",
        vatAmount: invoiceDto.vatAmount || "",
        totalAmountIncVat: invoiceDto.totalAmountIncVat || "",
        currencyCode: invoiceDto.currencyCode || "",
        bankAccount: invoiceDto.bankAccount || "",
        variableSymbol: invoiceDto.variableSymbol || "",
        paymentMethod: invoiceDto.paymentMethod || "",
        status: invoiceDto.status || "",
    };
};

/**
 * Mapuje formulář a položky na CreateInvoiceDTO
 * @param {Object} formData
 * @param {Array} items
 * @returns {Object}
 */
export const mapFormToCreateInvoiceDto = (formData, items) => {
    return {
        issueDate: formData.issueDate,
        taxableSupplyDate: formData.taxableSupplyDate,
        dueDate: formData.dueDate,
        supplierId: parseInt(formData.supplierId),
        customerId: parseInt(formData.customerId),
        totalAmountExVat: parseFloat(formData.totalAmountExVat) || 0,
        vatAmount: parseFloat(formData.vatAmount) || 0,
        totalAmountIncVat: parseFloat(formData.totalAmountIncVat) || 0,
        currencyCode: formData.currencyCode || "Kč",
        bankAccount: formData.bankAccount,
        variableSymbol: formData.variableSymbol,
        paymentMethod: formData.paymentMethod,
        status: formData.status || "",
        items: items.map((item) => ({
            description: item.description,
            quantity: parseInt(item.quantity) || 1,
            unitPrice: parseFloat(item.unitPrice) || 0,
            vatRate: parseFloat(item.vatRate) || 0,
            totalPrice: parseFloat(item.totalPrice) || 0,
        })),
    };
};

/**
 * Mapuje formulář na UpdateInvoiceDTO
 * @param {Object} formData
 * @returns {Object}
 */
export const mapFormToUpdateInvoiceDto = (formData) => {
    const dto = {};

    if (formData.invoiceNumber) dto.invoiceNumber = formData.invoiceNumber;
    if (formData.issueDate) dto.issueDate = formData.issueDate;
    if (formData.taxableSupplyDate) dto.taxableSupplyDate = formData.taxableSupplyDate;
    if (formData.dueDate) dto.dueDate = formData.dueDate;
    if (formData.status) dto.status = formData.status;
    if (formData.totalAmountExVat)
        dto.totalAmountExVat = parseFloat(formData.totalAmountExVat);
    if (formData.vatAmount) dto.vatAmount = parseFloat(formData.vatAmount);
    if (formData.totalAmountIncVat)
        dto.totalAmountIncVat = parseFloat(formData.totalAmountIncVat);
    if (formData.currencyCode) dto.currencyCode = formData.currencyCode;
    if (formData.bankAccount) dto.bankAccount = formData.bankAccount;
    if (formData.variableSymbol) dto.variableSymbol = formData.variableSymbol;
    if (formData.paymentMethod) dto.paymentMethod = formData.paymentMethod;

    return dto;
};

/**
 * Mapuje položku faktury na InvoiceItemDTO
 * @param {Object} item
 * @returns {Object}
 */
export const mapFormItemToInvoiceItemDto = (item) => {
    return {
        description: item.description,
        quantity: parseInt(item.quantity) || 1,
        unitPrice: parseFloat(item.unitPrice) || 0,
        vatRate: parseFloat(item.vatRate) || 0,
        totalPrice: parseFloat(item.totalPrice) || 0

    };
};
